src/main/java/org/candlepin/subscriptions/event/EventController.java

```diff
@@ -128,63 +146,124 @@ public Optional<OffsetDateTime> findFirstEventTimestampInRange(
+  private ServiceInstancesResult parseServiceInstancesResult(List<String> eventJsonList) {
     ServiceInstancesResult result = new ServiceInstancesResult();
-    eventJsonList.forEach(
-        eventJson -> {
-          try {
-            BaseEvent baseEvent = objectMapper.readValue(eventJson, BaseEvent.class);
-            if (!EXCLUDE_LOG_FOR_EVENT_SOURCES.contains(baseEvent.getEventSource())) {
-              log.info("Event processing in batch: " + baseEvent);
-            }
-
-            if (StringUtils.hasText(baseEvent.getOrgId())) {
-              log.debug(
-                  "Ensuring orgId={} has been set up for syncing/reporting.", baseEvent.getOrgId());
-              ensureOptIn(baseEvent.getOrgId());
-            }
-
-            if (baseEvent instanceof Event) {
-              result.addEvent((Event) baseEvent);
-            } else if (baseEvent instanceof CleanUpEvent) {
-              CleanUpEvent cleanUpEvent = (CleanUpEvent) baseEvent;
-              log.debug("Processing clean up event for: " + cleanUpEvent);
-              result.addCleanUpEvent(cleanUpEvent);
-            }
-
-          } catch (Exception e) {
-            log.warn(
-                String.format(
-                    "Issue found %s for the service instance json skipping to next ",
-                    e.getCause()));
-          }
-        });
+    Map<String, Integer> eventIndexMap = mapEventsToBatchIndex(eventJsonList);
+    for (Entry<String, Integer> eventIndex : eventIndexMap.entrySet()) {
+      try {
+        BaseEvent baseEvent = objectMapper.readValue(eventIndex.getKey(), BaseEvent.class);
+        if (!EXCLUDE_LOG_FOR_EVENT_SOURCES.contains(baseEvent.getEventSource())) {
+          log.info("Event processing in batch: " + baseEvent);
+        }
+        if (StringUtils.hasText(baseEvent.getOrgId())) {
+          log.debug(
+              "Ensuring orgId={} has been set up for syncing/reporting.", baseEvent.getOrgId());
+          ensureOptIn(baseEvent.getOrgId());
+        }
+
+        if (baseEvent instanceof Event eventToSave) {
+          result.addEvent(eventToSave, eventIndex.getValue());
+        } else if (baseEvent instanceof CleanUpEvent cleanUpEvent) {
+          log.debug("Processing clean up event for: " + cleanUpEvent);
+          result.addCleanUpEvent(cleanUpEvent);
+        }
+
+      } catch (Exception e) {
+        log.warn(
+            String.format(
+                "Issue found %s for the service instance json skipping to next ", e.getMessage()));
+        if (result.failedOnIndex.isEmpty()) {
+          result.setFailedOnIndex(eventIndex.getValue());
+        }
+        break;
+      }
+    }
+    return result;
+  }
+
+  /**
+   * Deduplicate eventJsonList while preserving indexes of events in batch
+   *
+   * @param eventJsonList
+   */
+  private Map<String, Integer> mapEventsToBatchIndex(List<String> eventJsonList) {
+    var result = new HashMap<String, Integer>();
```

Comment:
@kartikshahc I don't think I could get it to work with replacing the list with a set because I wanted to skip the records that failed to parse and we need to use the index for that. This method deduplicates the data upfront instead and keeps track of what index it was at for retrying.